### PR TITLE
*: pass sql, plan digest down to KV request

### PR DESCRIPTION
**Pull request by crazycs520**
Track issue: https://github.com/pingcap/tidb/issues/24875

Related PR: https://github.com/pingcap/tidb/pull/24380

### What problem does this PR solve?

This PR use to implement the part of TopSQL feature.

For TiKV to collection resource by SQL statements, TiDB need pass the `sql digest` and `plan digest` down to TiKV.

### What is changed and how it works?

1. Encode the `sql digest` and `plan digest` to `[]byte` and the set on `kv.Context.ResourceGroupTag`.

### Related PR

- https://github.com/pingcap/tipb/pull/223 add resource group tag definition
- https://github.com/pingcap/parser/pull/1231 Add Digest struct to get bytes of digest

### Check List 

Tests 

- Unit test

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

- No release note

### CODE REVIEW

**Review by MyonKeminta — COMMENTED** [review]
I think it doesn't look so good that the pessimistic lock requests use a different way to set the resource group tag. I think there should be a better way, but it's ok for now since we don't have too much time.

File: config/config.go [hunk 1]
```diff
@@ -137,6 +137,7 @@ type Config struct {
 	DelayCleanTableLock uint64      `toml:"delay-clean-table-lock" json:"delay-clean-table-lock"`
 	SplitRegionMaxNum   uint64      `toml:"split-region-max-num" json:"split-region-max-num"`
 	StmtSummary         StmtSummary `toml:"stmt-summary" json:"stmt-summary"`
+	TopStmt             TopSQL      `toml:"top-sql" json:"top-sql"`
```
[the remark this answers is not in the document]
Comment (crazycs520) [reply config/config.go]: already rename it.

File: executor/adapter.go [hunk 2]
```diff
@@ -919,11 +921,11 @@ func (a *ExecStmt) LogSlowQuery(txnTS uint64, succ bool, hasMoreResults bool) {
 	statsInfos := plannercore.GetStatsInfo(a.Plan)
 	memMax := sessVars.StmtCtx.MemTracker.MaxConsumed()
 	diskMax := sessVars.StmtCtx.DiskTracker.MaxConsumed()
-	_, planDigest := getPlanDigest(a.Ctx, a.Plan)
+	planDigest := getPlanDigest(a.Ctx, a.Plan)
```
Comment (tiancaiamao) [inline executor/adapter.go:924]: Duplicated with line 336?

File: executor/update.go [hunk 3]
```diff
@@ -263,6 +264,12 @@ func (e *UpdateExec) updateRows(ctx context.Context) (int, error) {
 				txn.GetSnapshot().SetOption(kv.CollectRuntimeStats, e.stats.SnapshotRuntimeStats)
 			}
 		}
+		if config.TopSQLEnabled() {
+			txn, err := e.ctx.Txn(false)
```
Comment (tiancaiamao) [inline executor/update.go:268]: It should be `e.ctx.Txn(true)` here.
To call  `SetOption()`, the underlying `kv.Transaction` must be actived.

File: util/resourcegrouptag/resource_group_tag_test.go [hunk 4]
```diff
@@ -29,83 +36,92 @@ func TestT(t *testing.T) {
...
+	}
+	return res
+}
 
-	tag = []byte("\x01\x01\x02")
-	_, err = DecodeResourceGroupTag(tag)
-	c.Assert(err, NotNil)
+func genDigest(str string) []byte {
+	hasher := sha256.New()
+	hasher.Write([]byte(str))
```
Comment (tiancaiamao) [inline util/resourcegrouptag/resource_group_tag_test.go:78]:
```
import "github.com/pingcap/tidb/util/hack

hasher.Write(hack.Slice(str))
```

File: sessionctx/stmtctx/stmtctx.go [hunk 5]
```diff
@@ -229,28 +232,40 @@ func (sc *StatementContext) ResetStmtCache() {
...
 	return sc.planNormalized, sc.planDigest
 }
 
+// GetResourceGroupTag gets the resource group of the statement.
+func (sc *StatementContext) GetResourceGroupTag() []byte {
+	if len(sc.resourceGroupTag) > 0 {
+		return sc.resourceGroupTag
+	}
+	_, sqlDigest := sc.SQLDigest()
+	sc.resourceGroupTag = resourcegrouptag.EncodeResourceGroupTag(sqlDigest, sc.planDigest)
```
Comment (lysu) [inline sessionctx/stmtctx/stmtctx.go:260]: could we just marshal once for an stmt? it seems N kv requests in a SQL should be same resourceTagValue

**Comment by lysu** [comment]
https://github.com/pingcap/tidb/blob/3f0c2a36e45a2257fbdd72e7fba90ece7c091159/store/driver/txn/txn_driver.go#L96

will be used in insert/replace code path (`prefetchUniqueIndices`..) , maybe also need `setResourceGroupTagForTxn(stmtCtx, snapshot)`, because it's new snapshot

File: sessionctx/stmtctx/stmtctx.go (hunk 5 above)
Comment (crazycs520) [reply sessionctx/stmtctx/stmtctx.go:260]: Currently, already cache the marshal result in `sc.resourceGroupTag`. Normally it should only marshal once.

File: executor/adapter.go (hunk 2 above)
Comment (crazycs520) [reply executor/adapter.go:924]: Line 336 uses to initialize the plan digest, and it is already cache in `StatementContext`

**Comment by crazycs520** [comment]
> https://github.com/pingcap/tidb/blob/3f0c2a36e45a2257fbdd72e7fba90ece7c091159/store/driver/txn/txn_driver.go#L96
> 
> will be used in insert/replace code path (`prefetchUniqueIndices`..) , maybe also need `setResourceGroupTagForTxn(stmtCtx, snapshot)`, because it's new snapshot

@lysu , Already call `setResourceGroupTagForTxn` for `replace` SQL in [replace.go#227](https://github.com/pingcap/tidb/pull/24854/files#diff-64a95f3c9846c8ce9e0aea454c5ac1742f39fc8cd4f6d85d609564a32b5ee2bdR227)

File: util/resourcegrouptag/resource_group_tag_test.go (hunk 4 above)
Comment (crazycs520) [reply util/resourcegrouptag/resource_group_tag_test.go]: nice, done.

**Review by lysu — APPROVED** [review]

File: sessionctx/stmtctx/stmtctx.go [hunk 6]
```diff
@@ -229,28 +232,45 @@ func (sc *StatementContext) ResetStmtCache() {
...
+	tag, _ := sc.resourceGroupTag.Load().([]byte)
+	if len(tag) > 0 {
+		return tag
+	}
+	normalized, sqlDigest := sc.SQLDigest()
+	if len(normalized) == 0 {
+		return nil
+	}
+	tag = resourcegrouptag.EncodeResourceGroupTag(sqlDigest, sc.planDigest)
+	sc.resourceGroupTag.Store(tag)
```
Comment (breeswish) [inline sessionctx/stmtctx/stmtctx.go:265]: What about concurrently calling this function?

File: util/deadlockhistory/deadlock_history.go [hunk 7]
```diff
@@ -183,7 +183,7 @@ func ErrDeadlockToDeadlockRecord(dl *tikverr.ErrDeadlock) *DeadlockRecord {
 		}
 		waitChain = append(waitChain, WaitChainItem{
 			TryLockTxn:     rawItem.Txn,
-			SQLDigest:      sqlDigest,
+			SQLDigest:      hex.EncodeToString(sqlDigest),
```
Comment (breeswish) [inline util/deadlockhistory/deadlock_history.go:186]: Is it better to be `NewDigest(sqlDigest).String()`?

File: sessionctx/stmtctx/stmtctx.go (hunk 6 above)
Comment (crazycs520) [reply sessionctx/stmtctx/stmtctx.go:265]: Concurrently calling doesn't affect correctness.

File: util/deadlockhistory/deadlock_history.go (hunk 7 above)
Comment (crazycs520) [reply util/deadlockhistory/deadlock_history.go:186]: Yes,  use `NewDigest(sqlDigest).String()` is better.

**Pull request merged**